[Merged by Bors] - feat(MeasureTheory): method of exhaustion

RemyDegenne opened the pull request:
If `μ, ν` are two measures with `ν` s-finite, then there exists a set `s` such that `μ` is sigma-finite on `s`, and for all sets `t ⊆ sᶜ`, either `ν t = 0` or `μ t = ∞`. The technique used to obtain this result is what Halmos calls the "method of exhaustion".

We use this to prove that if `μ ≪ ν` and `ν` is s-finite, then `μ` is s-finite.

This PR also moves the definition of `sigmaFiniteSet` to the new file and refactors it with the new definition `sigmaFiniteSetWRT`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks, and is labeled with `awaiting-review`.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sgouezel commented on Mathlib/MeasureTheory/Decomposition/Exhaustion.lean:
```diff
@@ -0,0 +1,207 @@
+/-!
+# Method of exhaustion
+
+If `μ, ν` are two measures with `ν` finite, then there exists a set `s` such that
+`μ` is sigma-finite on `s`, and for all sets `t ⊆ sᶜ`, either `ν t = 0` or `μ t = ∞`.
```
I would expect that this statement generalizes readily when nu is S-finite (because it is then equivalent to a finite measure). What about defining `sigmaFiniteSetWRT` without any assumptions on nu as: if there exists a measurable set s such that (mu.restrict s is sigma-finite and for all t in s^c, either nu t = 0 or mu t = \infty), then pick such a set s, else pick the empty set. Then prove that for finite measures, it does what it should do (following exactly your proof), and then deduce the same for S-finite nu?

RemyDegenne commented:
@sgouezel I generalized the definition of `sigmaFiniteSetWRT` in the way you suggested.
I have added another change to the PR: I changed the definition of `sigmaFiniteSet` to use the new `sigmaFiniteSetWRT` and moved the whole thing to the new Exhaustion file.

RemyDegenne commented:
The import change bot is saying "Too many changes (249)!" but it looks like that simply means that 249 files now import the new Exhaustion file (since it is imported by `WithDensity`).

semorrison commented:
> The import change bot is saying "Too many changes (249)!" but it looks like that simply means that 249 files now import the new Exhaustion file (since it is imported by `WithDensity`).

Oh, we should definitely change that bot so the message is more clearly a failure to display the differences, target than objecting to the PR!

@adomani?

sgouezel commented on Mathlib/MeasureTheory/Decomposition/Exhaustion.lean:
```diff
@@ -0,0 +1,335 @@
+    MeasurableSet (μ.sigmaFiniteSetGE ν n) :=
+  (exists_isSigmaFiniteSet_measure_ge μ ν n).choose_spec.1
+
+instance (n : ℕ) : SigmaFinite (μ.restrict (μ.sigmaFiniteSetGE ν n)) :=
```
Since this construction is purely local to this file, I wouldn't register it as a global instance, to avoid cluttering instance search. Can you make it a local instance instead?

sgouezel commented on Mathlib/MeasureTheory/Decomposition/Exhaustion.lean:
```diff
@@ -0,0 +1,335 @@
+/-- Let `C` be the supremum of `ν s` over all measurable sets `s` such that `μ.restrict s` is
+sigma-finite. `C` is finite since `ν` is a finite measure. Then there exists a measurable set `t`
+with `μ.restrict t` sigma-finite such that `ν t ≥ C - 1/n`. -/
+lemma exists_isSigmaFiniteSet_measure_ge (μ ν : Measure α) [IsFiniteMeasure ν] (n : ℕ) :
```
The assumptions `(μ ν : Measure α) [IsFiniteMeasure ν]` show up in this lemma and all the next ones. Can you make them into a `variable` line?

RemyDegenne replied:
I'd rather like to keep all explicit arguments with the lemma definitions.

sgouezel commented on Mathlib/MeasureTheory/Decomposition/Exhaustion.lean:
```diff
@@ -0,0 +1,335 @@
+    MeasurableSet (μ.sigmaFiniteSetWRT' ν) :=
+  MeasurableSet.iUnion (measurableSet_sigmaFiniteSetGE _ _)
+
+instance [IsFiniteMeasure ν] : SigmaFinite (μ.restrict (μ.sigmaFiniteSetWRT' ν)) := by
```
ditto, make it a local instance?

sgouezel commented on Mathlib/MeasureTheory/Decomposition/Exhaustion.lean:
```diff
@@ -0,0 +1,335 @@
+
+/-- For all measurable sets `s` in `(μ.sigmaFiniteSetWRT ν)ᶜ`, if `ν s ≠ 0` then `μ s = ∞`. -/
+lemma measure_eq_top_of_subset_compl_sigmaFiniteSetWRT'
+    (hs : MeasurableSet s) (hs_subset : s ⊆ (μ.sigmaFiniteSetWRT' ν)ᶜ) (hνs : ν s ≠ 0) :
```
Once you've proved this one, can you drop the `MeasurableSet` assumption, by arguing that the measure is the inf of the measure of measurable sets containing them (which you can take inside `(μ.sigmaFiniteSetWRT' ν)ᶜ` as it's measurable)?

sgouezel commented on Mathlib/MeasureTheory/Decomposition/Exhaustion.lean:
```diff
@@ -0,0 +1,335 @@
+section SFinite
+
+lemma measure_eq_top_of_subset_compl_sigmaFiniteSetWRT [SFinite ν]
+    (hs : MeasurableSet s) (hs_subset : s ⊆ (μ.sigmaFiniteSetWRT ν)ᶜ) (hνs : ν s ≠ 0) :
```
If my comment above works fine, then you should be able to remove `hs` here, and then the lemma ` measure_eq_top_of_absolutelyContinuous_of_subset_compl_sigmaFiniteSetWRT` becomes redundant, right?

adomani commented:
Oh, I had not thought of the possible interpretation that the bot was rejecting the PR!  I introduced the limit on number of changed files, since otherwise, if the message is too long, then GitHub will just fail to post it.

I'll update the message, but still leave on the stop-gag, to avoid corrupting the `PR summary` message.

RemyDegenne commented:
I don't know how to turn instances local. When I add the `local` keyword in front of them, the theorem vs def linter complains.

adomani commented:
This is exactly the reason the linter complains: as far as I know, there is no way to make an instance local.  There is a lean4 issue about this.  I'll look for it and edit this comment.

RemyDegenne commented:
Thanks for the info! I'll turn them into lemmas.

sgouezel reviewed (commented):
bors d+
Thanks!

RemyDegenne commented:
bors r+
